### ISSUE TITLE
Trainer: add text trainer for veomni

### ISSUE THREAD

**Issue by Fazziekey** [ByteDance-Seed/VeOmni#326]
### Feature request

# [Feature Request] Add Trainer Support for VeOmni

## Overview

This feature request proposes adding a modular Trainer framework to VeOmni to provide a flexible and extensible training infrastructure. The Trainer should support omni-modality training scenarios with a callback system for easy customization.

## Background

VeOmni currently needs a more flexible and extensible training framework. A well-designed Trainer module would provide:

1. **Modular Design**: Abstract base class architecture that's easy to extend
2. **Simple Callback System**: Minimal callback interface for evaluation and logging (user-initiated, not automatic)
3. **Distributed Training Support**: Built-in support for data parallelism, tensor parallelism, pipeline parallelism, etc.
4. **Flexible Batching**: Support for dynamic batching, gradient accumulation, and other advanced features

### Core Components

```
BaseTrainer (abstract base class)
    └── Trainer (omni-modality training implementation)

BaseCallback (callback base class)
    ├── TrainerCallback (general training metrics)
    └── [Custom callbacks for specific use cases]
```

### Key Features

- **Distributed Training Initialization**: Automatic handling of DP/TP/PP/EP/CP parallel strategies
- **Model Parallelization**: Support for FSDP, gradient checkpointing, activation offloading
- **Data Loading**: Integration with dynamic batching, packing, sequence sharding
- **Optimizer and Scheduler**: Flexible optimizer and learning rate scheduler construction
- **Checkpoint Management**: Distributed checkpoint saving and loading
- **Metrics Collection**: Built-in environment metrics collection and WandB integration
- **Simple Callback System**: Minimal callback interface with only `evaluate()` and `on_log()` methods, requiring explicit user invocation

### Core Methods

The Trainer should provide the following key methods:

- `fit()`: Main training loop
- `train_step()`: Single training step execution
- `forward_backward_step()`: Forward and backward pass
- `pre_forward()`: Data preprocessing before forward pass
- `batch_generator()`: Batch processing generator
- `add_callback()`: Add custom callbacks

### Proposed Structure

```
veomni/trainer/
├── __init__.py          # Public API exports
├── base.py              # BaseTrainer abstract base class
├── trainer.py           # VeOmniTrainer implementation
└── callbacks/
    ├── __init__.py
    ├── base.py          # BaseCallback
    └── trainer_callback.py  # Default trainer callback
```

### Callback Design Philosophy

The callback system is intentionally **simple and explicit**:

- **Only two methods**: `evaluate()` and `on_log()`
- **User-initiated**: Callbacks are called explicitly by the user, not automatically triggered by the trainer
- **No lifecycle hooks**: Unlike frameworks like [PyTorch Lightning](https://github.com/Lightning-AI/pytorch-lightning) which have complex callback systems with hooks like `on_fit_start()`, `on_train_epoch_start()`, `on_train_batch_start()`, etc., VeOmni's callback system avoids automatic lifecycle management
- **Explicit control**: Users have full control over when callbacks are invoked

This design choice prioritizes:
- **Simplicity**: Easy to understand and use
- **Explicitness**: Clear control flow without hidden automatic behaviors
- **Flexibility**: Users can implement their own training loop logic without fighting against automatic callback triggers

### Implementation Steps

1. **Create BaseTrainer**: Abstract base class defining the trainer interface
2. **Implement Trainer**: Concrete implementation for omni-modality training
3. **Add Simple Callback System**: Base callback class with only `evaluate()` and `on_log()` methods
4. **Integrate with Existing Components**: 
   - Model building interfaces
   - Data loading interfaces
   - Checkpoint management
   - Distributed training setup

### Basic Usage

```python
from veomni.trainer import Trainer
from veomni.utils.arguments import parse_args, Arguments

# Parse arguments
args = parse_args(Arguments)

# Create trainer
trainer = Trainer(args)

# Start training
trainer.fit()
```

### Custom Callback (User-Initiated)

```python
from veomni.trainer import Trainer
from veomni.trainer.callbacks import BaseCallback

class CustomCallback(BaseCallback):
    def evaluate(self, trainer, **kwargs):
        # Custom evaluation logic
        # This is called explicitly by the user, e.g., trainer._evaluate()
        pass
    
    def on_log(self, trainer, **kwargs):
        # Custom logging logic
        # This is called explicitly by the user, e.g., trainer._log()
        return {}

trainer = Trainer(args)
trainer.add_callback(CustomCallback())

# Callbacks are invoked explicitly by the user
for epoch in range(num_epochs):
    for step in range(num_steps):
        trainer.train_step(data_iterator)
        
        # User explicitly calls evaluation callback
        if step % eval_steps == 0:
            trainer._evaluate()  # Calls callback.evaluate() for all callbacks
        
        # User explicitly calls logging callback
        if step % log_steps == 0:
            trainer._log()  # Calls callback.on_log() for all callbacks
```

**Note**: Unlike PyTorch Lightning's automatic callback system with hooks like `on_fit_start()`, `on_train_epoch_start()`, etc., VeOmni's callbacks require explicit invocation. This gives users full control over when callbacks are executed.

### Omni-Modality Training Example

```python
from veomni.trainer import BaseTrainer

class VeOmniTrainer(BaseTrainer):
    def pre_forward(self, input_dict, labels=None):
        # Process omni-modality inputs (vision, text, etc.)
        vision_inputs = input_dict.get("pixel_values")
        text_inputs = input_dict.get("input_ids")
        # ... omni-modality preprocessing logic
        return processed_inputs
    
    def forward_backward_step(self, input_dict, labels=None):
        # Omni-modality forward pass
        outputs = self.model(**input_dict)
        loss = outputs.loss
        loss.backward()
        return loss
```

## Benefits

1. **Extensibility**: Abstract base class makes it easy to extend with new trainer types
2. **Modularity**: Decoupled components for easier maintenance and testing
3. **Simple Callback Mechanism**: Minimal callback interface (evaluate and log only) with explicit user control, avoiding complex automatic lifecycle hooks
4. **Distributed Support**: Built-in comprehensive distributed training support
5. **Checkpoint Management**: Unified checkpoint save/load interface
6. **Metrics Collection**: Built-in performance metrics collection and visualization

## Compatibility Considerations

- **Model Interface**: Ensure VeOmni models conform to Trainer interface requirements
- **Data Format**: Adapt to VeOmni's data loading format
- **Checkpoint Format**: May need to adapt checkpoint save/load format
- **Evaluation Flow**: Implement omni-modality specific evaluation logic

## Implementation Phases

1. **Phase 1**: Implement base Trainer with basic training flow support
2. **Phase 2**: Implement omni-modality specific Trainer subclass
3. **Phase 3**: Add omni-modality evaluation callbacks and visualization
4. **Phase 4**: Performance optimization and advanced features

### Why Simple Callbacks?

Unlike frameworks such as [PyTorch Lightning](https://github.com/Lightning-AI/pytorch-lightning) which provide extensive callback hooks (`on_fit_start()`, `on_train_epoch_start()`, `on_train_batch_start()`, `on_train_batch_end()`, etc.), VeOmni's callback system is intentionally minimal:

1. **Explicit Control**: Users explicitly call `trainer._evaluate()` and `trainer._log()` when needed, making the control flow clear and predictable
2. **Simplicity**: Only two methods (`evaluate()` and `on_log()`) are easier to understand and implement
3. **No Hidden Behavior**: Avoids automatic callback triggers that can make debugging difficult
4. **Flexibility**: Users can implement custom training loops without fighting against automatic callback systems

### Callback Interface

```python
class BaseCallback(ABC):
    @abstractmethod
    def evaluate(self, trainer, **kwargs) -> None:
        """Called explicitly via trainer._evaluate()"""
        pass
    
    @abstractmethod
    def on_log(self, trainer, **kwargs) -> Dict[str, Any]:
        """Called explicitly via trainer._log()"""
        pass
```

## Related Components

The Trainer should integrate with:
- Model building infrastructure
- Data loading pipeline
- Checkpoint management system
- Distributed training utilities
- Metrics collection system

## Labels

- `enhancement`
- `roadmap`
- `training`

## References

- Related Issue: #268 (VeOmni v0.2 Release Plan)
- Existing VeOmni training workflows

### Motivation

-

### Your contribution

-

**Comment by Coach257** [comment]
similar implementation in verl: https://github.com/volcengine/verl/blob/7ba518e1f31b0991ca42cfaf7328a4131e99dd55/verl/workers/engine/veomni/transformer_impl.py#L50

**Comment by Luosuu** [comment]
To be honest, I think at least we should not break the current trainers under `tasks/` so that users can still use the linear style training script. The trainer module looks scary to me since it looks hard to debug.

I totally see the value of trainer module so that external library can register VeOmni as a standalone training module but we should not replace the current `tasks/` trainers.

Related discussions:
https://www.sicpers.info/2018/03/why-inheritance-never-made-any-sense/
https://www.reddit.com/r/rust/comments/1q3u3wj/whats_wrong_with_subtypes_and_inheritance/

**Comment by Luosuu** [comment]
Reference trianer: https://github.com/allenai/OLMo-core/blob/main/src/olmo_core/train/trainer.py

**Comment by Coach257** [comment]
#458
